airbyte-integrations/connectors/destination-s3/src/test-integration/kotlin/io/airbyte/integrations/destination/s3/S3StorageOperationsIntegrationTest.kt

```diff
@@ -0,0 +1,286 @@
+        uploadTestRecords(namespace, fullObjectPath, 12, 1)
+        // write 10 files with genId 2
+        uploadTestRecords(namespace, fullObjectPath, 10, 2)
+        // write 5 files with genId 3
+        uploadTestRecords(namespace, fullObjectPath, 5, 3)
+
+        val existingObjects =
+            s3StorageOperations.listExistingObjects(
+                namespace,
+                streamName,
+                s3DestinationConfig.bucketPath!!,
+                outputFormat
+            )
+        assertEquals(27, existingObjects.size)
+    }
+
+    @ParameterizedTest
+    @MethodSource("listExistingObjectTestDataProvider")
+    fun testListExistingObjectsWithGenIdFiltered(
+        numberOfGens: Int,
+        includeNullGen: Boolean,
+        currentGen: Int
+    ) {
+        val namespace = null
+        val streamName = "testStream"
+        val outputFormat = "${s3DestinationConfig.bucketPath}/${s3DestinationConfig.pathFormat}"
+        val fullObjectPath =
+            s3StorageOperations.getBucketObjectPath(
+                namespace,
+                streamName,
+                DateTime.now(DateTimeZone.UTC),
+                outputFormat
+            )
+        var expectedNumberOfObjects = 0
+        if (numberOfGens != 0) {
+            for (i in 1..numberOfGens) {
+                val numberOfObjects =
+                    when (val randomNumber = random.nextInt(5)) {
+                        0 -> 1
+                        else -> randomNumber
+                    }
+                expectedNumberOfObjects += numberOfObjects
+                uploadTestRecords(namespace, fullObjectPath, numberOfObjects, i.toLong())
+            }
+        }
+
+        // Skip records to insert because the algorithm is
+        // tightly dependent on both lastModified and genId to be in the same sorted order
+        // This is used to assert non-existent case where currentGen gets decreased from highest Gen
+        if (currentGen > numberOfGens) {
+            val currentGenObjectSize = random.nextInt(5)
+            uploadTestRecords(namespace, fullObjectPath, currentGenObjectSize, currentGen.toLong())
+        }
+
+        val existingObjects =
+            s3StorageOperations.listExistingObjects(
+                namespace,
+                streamName,
+                s3DestinationConfig.bucketPath!!,
+                outputFormat,
+                currentGen.toLong()
+            )
+        assertEquals(expectedNumberOfObjects, existingObjects.size)
+    }
+
+    private fun createStringBuffer(strData: String): SerializableBuffer {
+        // Intentionally not implementing the methods to see the surface area needed for tests.
+        return object : SerializableBuffer {
+            @Deprecated("")
+            override fun accept(
+                record: AirbyteRecordMessage,
+                generationId: Long,
+                syncId: Long
+            ): Long {
+                TODO("Not yet implemented")
```

Comment:
nit: `throw NotImplementedException()` since we don't intend to ever implement these

Reply:
done